File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.py
class Solution:
    def check(self,positions,mid):
        curr=positions[0]
        ans=1
        for i in range(1,len(positions)):
            if positions[i]-curr>=mid:
                ans+=1
                curr=positions[i]
        return ans
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        n=len(position)
        l=0
        r=position[-1]-position[0]+1
        while l<r:
            mid=(l+r)//2
            if self.check(position,mid)>=m:
                l=mid+1
            else:
                r=mid
            #print(l,r)    
        return l-1
                
        

File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.cpp
class Solution {
public:
    
    int solve(vector<int>& positions,int count){
        int ans=1;
        //check difference
        int curr=positions[0];
        for(int i=1;i<positions.size();i++){
            if (positions[i]-curr>=count){
                ans++;
                curr=positions[i];
            }
        }
        return ans;
    }
    int maxDistance(vector<int>& positions, int m) {
        //sort the positions array
        //binary search the target value
        sort(positions.begin(),positions.end());
        int l=0;
        int n=positions.size();
        //max ans possible
        int r=(positions[n-1]-positions[0]);
        int res=0;
        while (l<=r){
            int mid=l+(r-l)/2;
            //if solve returns split more than the desired value we can increase the left
            if(solve(positions,mid)>=m){
                l=mid+1;
                res=mid;
            }
            //decrease the right
            else{
                r=mid-1;
            }
        }
        return res;
        
    }
};